File: src/universe/entity/entities/VehicleDebug.cpp
#include "VehicleDebug.h"
#include <imgui/imgui.h>
#include "VehicleEntity.h"

void VehicleDebug::show_imgui()
{
	std::vector<Machine*> closed;
	for(auto it = shown_machines.begin(); it != shown_machines.end();)
	{
		bool opened = true;
		(*it)->draw_imgui(&opened);
		if(opened)
		{
			it++;
		}
		else
		{
			it = shown_machines.erase(it);
		}
	}

	ImGui::SetNextWindowSize(ImVec2(600, 700), ImGuiCond_FirstUseEver);
	ImGui::Begin(("Vehicle (" + std::to_string(ent->get_uid()) + ")").c_str());

	if(ImGui::BeginTabBar("##tabs"))
	{
		if(ImGui::BeginTabItem("Parts"))
		{
			parts_tab();
			ImGui::EndTabItem();
		}

		if(ImGui::BeginTabItem("Pieces"))
		{

			ImGui::EndTabItem();
		}

		if(ImGui::BeginTabItem("Plumbing"))
		{

			ImGui::EndTabItem();
		}

		if(ImGui::BeginTabItem("Wiring"))
		{

			ImGui::EndTabItem();
		}

		ImGui::EndTabBar();
	}
	ImGui::End();
}

void VehicleDebug::parts_tab()
{
	// We show all parts in the vehicle
	// Show all machines
	for(Part* p : ent->vehicle->parts)
	{
		ImGui::PushID(p);
		std::string name = "Part (" + p->part_proto->get_asset_id() + ") " + std::to_string(p->id);
		if(ImGui::CollapsingHeader(name.c_str()))
		{
			ImGui::Button("Focus");
			ImGui::Text("Machines: ");
			for(auto pair : p->get_all_machines())
			{
				ImGui::PushID(pair.second);
				ImGui::Text("\t%s(%s)", pair.first.c_str(),
							pair.second->get_id().c_str());
				ImGui::SameLine();
				if(vector_contains(shown_machines, pair.second))
				{
					if (ImGui::Button("Hide Imgui"))
					{
						shown_machines.erase(std::find(shown_machines.begin(),
													   shown_machines.end(), pair.second));
					}
				}
				else
				{
					if(ImGui::Button("Show Imgui"))
					{
						shown_machines.push_back(pair.second);
					}
				}
				ImGui::SameLine();
				ImGui::Button("Show Wiring");
				ImGui::SameLine();
				ImGui::Button("Show Plumbing");
				ImGui::PopID();
			}

		}
		ImGui::PopID();
	}

}

File: res/core/entities/vehicle_debug.lua
local imgui = require("imgui")

local vehicle_debug = {}

vehicle_debug.shown_machines = {}

function vehicle_debug:draw_main()
	self:parts_tab()
end

function vehicle_debug:draw_machines()
	for ruid, path in pairs(self.shown_machines) do 
		local part = self.vehicle:get_part_by_id(path["part"])
		if part ~= nil then 	
			local machine = part:get_machine(path["machine"])
			if machine ~= nil then 
				if machine:draw_imgui(true) == false then 
					self.shown_machines[ruid] = nil
				end
			else
				self.shown_machines[ruid] = nil
			end
		else 
			self.shown_machines[ruid] = nil
		end
	end
end

function vehicle_debug:parts_tab()
	for _, part in ipairs(self.vehicle.parts) do
		imgui.push_id(part.id)
		local name = "Part (" .. part:get_part_proto():get_asset_id() .. ")" .. tostring(part.id)
		if imgui.collapsing_header(name) then
			imgui.text("Machines: ")
			for id, machine in part.machines:pairs() do 
				imgui.push_id(machine.runtime_uid)	
				imgui.text("\t" .. id .. "(" .. machine:get_id() .. ")")
				imgui.same_line()
				if self.shown_machines[machine.runtime_uid] ~= nil then 
					if imgui.button("Hide ImGui") then 
						self.shown_machines[machine.runtime_uid] = nil
					end
				else
					if imgui.button("Show ImGui") then
						local path = {}
						path["part"] = part.id
						path["machine"] = id
						self.shown_machines[machine.runtime_uid] = path
					end
				end
				imgui.pop_id()
			end
		end
		imgui.pop_id()
	end
end

---@param vehicle vehicle
function vehicle_debug:init(vehicle)
	self.vehicle = vehicle
	return self
end

return vehicle_debug
